src/xrpld/app/tx/detail/VaultClawback.cpp

```diff
@@ -60,179 +43,339 @@ VaultClawback::preclaim(PreclaimContext const& ctx)
+            return STAmount{share};
+
+        return STAmount{vaultAsset};
+    }();
+
+    // There is a special case that allows the VaultOwner to use clawback to
+    // burn shares when Vault assets total and available are zero, but
+    // shares remain. However, that case is handled in doApply() directly,
+    // so here we just enforce checks.
+    if (amount.asset() == share)
     {
-        auto const mpt = vaultAsset.get<MPTIssue>();
-        auto const mptIssue =
-            ctx.view.read(keylet::mptIssuance(mpt.getMptID()));
-        if (mptIssue == nullptr)
-            return tecOBJECT_NOT_FOUND;
-
-        std::uint32_t const issueFlags = mptIssue->getFieldU32(sfFlags);
-        if (!(issueFlags & lsfMPTCanClawback))
+        // Only the Vault Owner may clawback shares
+        if (account != vault->at(sfOwner))
+        {
+            JLOG(ctx.j.debug())
+                << "VaultClawback: only vault owner can clawback shares.";
+            return tecNO_PERMISSION;
+        }
+
+        auto const assetsTotal = vault->at(sfAssetsTotal);
+        auto const assetsAvailable = vault->at(sfAssetsAvailable);
+        auto const sharesTotal = sleShareIssuance->at(sfOutstandingAmount);
+
+        // Owner can clawback funds when the vault has shares but no assets
+        if (sharesTotal != 0 && (assetsTotal != 0 || assetsAvailable != 0))
         {
             JLOG(ctx.j.debug())
-                << "VaultClawback: cannot clawback MPT vault asset.";
+                << "VaultClawback: vault owner can clawback shares only"
+                   " when vault has no assets.";
             return tecNO_PERMISSION;
         }
+
+        // If amount is non-zero, the VaultOwner must burn all shares
+        if (amount != beast::zero)
+        {
+            Number const& sharesHeld = accountHolds(
+                ctx.view,
+                holder,
+                share,
+                FreezeHandling::fhIGNORE_FREEZE,
+                AuthHandling::ahIGNORE_AUTH,
+                ctx.j);
+
+            // The VaultOwner must burn all shares
+            if (amount != sharesHeld)
+            {
+                JLOG(ctx.j.debug())
+                    << "VaultClawback: vault owner must clawback all "
+                       "shares.";
+                return tecLIMIT_EXCEEDED;
+            }
+        }
+
+        return tesSUCCESS;
     }
-    else if (vaultAsset.holds<Issue>())
+
+    // The asset that is being clawed back is the vault asset
+    if (amount.asset() == vaultAsset)
     {
-        std::uint32_t const issuerFlags = issuer->getFieldU32(sfFlags);
-        if (!(issuerFlags & lsfAllowTrustLineClawback) ||
-            (issuerFlags & lsfNoFreeze))
+        // XRP cannot be clawed back
+        if (vaultAsset.native())
+        {
+            JLOG(ctx.j.debug()) << "VaultClawback: cannot clawback XRP.";
+            return tecNO_PERMISSION;
+        }
+
+        auto const issuer = ctx.view.read(keylet::account(account));
```

Comment:
Maybe `issuerSle` is a better name? `issuer` would normally refer to `AccountID`.

Reply:
I moved this to the block where it's actually used. THis lookup is not necessary for MPTs.